src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/SqlDataSourceBuilder.cs

```diff
@@ -42,11 +46,67 @@ public SqlDataSourceBuilder(string connectionString = null, SqlCredential creden
         /// </summary>
         public SqlDataSource Build()
         {
-            return new UnpooledDataSource(
-                ConnectionStringBuilder,
-                Credential,
-                UserCertificateValidationCallback,
-                ClientCertificatesCallback);
+            if (ConnectionStringBuilder.Pooling)
+            {
+                //TODO: pool group layer
+
+                DbConnectionPoolGroupOptions poolGroupOptions = new DbConnectionPoolGroupOptions(
+                    ConnectionStringBuilder.IntegratedSecurity,
+                    ConnectionStringBuilder.MinPoolSize,
+                    ConnectionStringBuilder.MaxPoolSize,
+                    //TODO: carry over connect timeout conversion logic from SqlConnectionFactory? if not, don't need an extra allocation for this object, just use connection string builder
+                    ConnectionStringBuilder.ConnectTimeout,
+                    ConnectionStringBuilder.LoadBalanceTimeout,
+                    ConnectionStringBuilder.Enlist);
+
+                //TODO: evaluate app context switch for concurrency limit
```

Comment:
We haven't taken a hard stance on this yet, but in general we want to enable new functionality by default and provide a mechanism for opt-out (using appropriate semver). Otherwise it will be practically impossible to drive adoption.

One motivation behind the SqlDataSourceBuilder pattern we're introducing is that we can also expose an ExperimentalSqlDataSourceBuilder which may break contract often and grants access to preview experimental features. As features are proven, they can be ported over to the base SqlDataSourceBuilder and default enabled with option to opt out.

Reply:
That makes sense. So the app context switch would essentially be the maximum "functionality version" available for use, with the default value being to enable all functionality in the SqlClient version? There's certainly prior art for this - Azure has its API versions, and Cloudflare do something similar with [compatibility dates](https://developers.cloudflare.com/workers/configuration/compatibility-dates). I don't think they enable functionality by default, but they're otherwise-versionless cloud services and SqlClient's a versioned library, which gives us that latitude; I've not got any objections to doing that.

I'm in two minds on whether the version should actually be the master, or whether we should have a collection of feature flags which can be independently enabled/disabled and enabled in bulk by setting the version. It'll probably depend on how many/which features we put behind the switch - testing those would scale factorially...

Reply:
Added an issue for this: #2678

Reply:
Agreed, we'll need to limit the number of flags we use and plan to phase them out over time.